src/NUnitFramework/tests/Constraints/CollectionEquivalentConstraintTests.cs

```diff
@@ -152,8 +152,19 @@ public void EquivalentHonorsUsing()
                 .Using<string>((x, y) => StringUtil.Compare(x, y, true))
                 .ApplyTo(set2).IsSuccess);
         }
+#if (NET_4_0 || NET_4_5) && !PORTABLE
 
-#if ( NET_4_0 || NET_4_5 ) && !PORTABLE
+        [Test]
+        public void EquivalentHonorsUsingWhenCollectionsAreOfDifferentTypes()
+        {
+            ICollection set1 = new SimpleObjectCollection("1", "2", "3");
+            ICollection set2 = new SimpleObjectCollection(1, 2, 3);
+
+            Assert.That(set2, Is.EquivalentTo(set1).Using<int, string>((i, s) => i.ToString() == s));
+        }
+#endif
```

Comment:
"I don't want users worrying about which overloads work for which constraints"
I agree.

What I'm proposing is to remove the implementation of Unique<TExpected, TActual>(lambda) out of the base class to three of subclasses. In this way it would only be available on the constraints where it makes sense.

`Is.Unique` is slightly different in what it's testing than the other constraint. I think it needs a generic method with the same _signature_, but the meaning of the arguments is different. Consider:

```csharp
        internal class Widget
        {
            public int Id { get; set; }
            public string Name { get; set; }
        }

        [Test]
        public void UniqueHonorsUsingByProperty()
        {

            ICollection set1 = new SimpleObjectCollection(
                new Widget()
            {
                Id = 1,
                Name = "Widget 1",
            }, new Widget()
            {
                Id = 2,
                Name= "Widget 2"
            }, new Widget()
            {
                Id = 3,
                Name = "Widget 3"
            });

            Assert.That(set1, Is.Unique(set1).Using<Widget, int>(w => w.Id));
        }
```

This would read "Assert that widgets are unique by their ids." 

There's an alternative way to implement this for Unique that only requires a single generic argument to Using:

```csharp
        [Test]
        public void UniqueHonorsUsingWithComparison()
        {

            ICollection set1 = new SimpleObjectCollection(
                new Widget()
                {
                    Id = 1,
                    Name = "Widget 1",
                }, new Widget()
                {
                    Id = 2,
                    Name = "Widget 2"
                }, new Widget()
                {
                    Id = 3,
                    Name = "Widget 3"
                });

            Assert.That(set1, Is.Unique(set1).Using<Widget>((left, right) => left.Id == right.Id));
        }
```

This latter signature could be supported in all 4 constraints with the same implementation and could be in the base class.